shell/platform/fuchsia/flutter/software_surface_producer.cc

```diff
@@ -0,0 +1,236 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "software_surface_producer.h"
+
+#include <lib/async/cpp/task.h>
+#include <lib/async/default.h>
+#include <lib/fdio/directory.h>
+#include <lib/zx/process.h>
+
+#include <algorithm>  // Foor std::remove_if
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "flutter/fml/trace_event.h"
+
+namespace flutter_runner {
+
+namespace {
+
+std::string GetCurrentProcessName() {
+  char name[ZX_MAX_NAME_LEN];
+  zx_status_t status =
+      zx::process::self()->get_property(ZX_PROP_NAME, name, sizeof(name));
+  return status == ZX_OK ? std::string(name) : std::string();
+}
+
+zx_koid_t GetCurrentProcessId() {
+  zx_info_handle_basic_t info;
+  zx_status_t status = zx::process::self()->get_info(
+      ZX_INFO_HANDLE_BASIC, &info, sizeof(info), nullptr, nullptr);
+  return status == ZX_OK ? info.koid : ZX_KOID_INVALID;
+}
+
+}  // namespace
+
+SoftwareSurfaceProducer::SoftwareSurfaceProducer(
+    scenic::Session* scenic_session)
+    : scenic_session_(scenic_session) {
+  zx_status_t status = fdio_service_connect(
+      "/svc/fuchsia.sysmem.Allocator",
+      sysmem_allocator_.NewRequest().TakeChannel().release());
+  sysmem_allocator_->SetDebugClientInfo(GetCurrentProcessName(),
+                                        GetCurrentProcessId());
+  FML_DCHECK(status != ZX_OK);
+
+  if (!scenic_session_) {
+    status = fdio_service_connect(
+        "/svc/fuchsia.ui.composition.Allocator",
+        flatland_allocator_.NewRequest().TakeChannel().release());
+    FML_DCHECK(status != ZX_OK);
+  }
+
+  valid_ = true;
+}
+
+SoftwareSurfaceProducer::~SoftwareSurfaceProducer() = default;
+
+std::unique_ptr<SurfaceProducerSurface>
+SoftwareSurfaceProducer::ProduceOffscreenSurface(const SkISize& size) {
+  FML_CHECK(valid_);
+
+  return CreateSurface(size);
+}
+
+std::unique_ptr<SurfaceProducerSurface> SoftwareSurfaceProducer::ProduceSurface(
+    const SkISize& size) {
+  TRACE_EVENT2("flutter", "SoftwareSurfacePool::ProduceSurface", "width",
+               size.width(), "height", size.height());
+  FML_CHECK(valid_);
+
+  std::unique_ptr<SurfaceProducerSurface> surface;
+  auto exact_match_it =
+      std::find_if(available_surfaces_.begin(), available_surfaces_.end(),
+                   [&size](const auto& surface) {
+                     return surface->IsValid() && surface->GetSize() == size;
+                   });
+  if (exact_match_it != available_surfaces_.end()) {
+    TRACE_EVENT_INSTANT0("flutter", "Exact match found");
+    surface = std::move(*exact_match_it);
+    available_surfaces_.erase(exact_match_it);
+  } else {
+    surface = CreateSurface(size);
+  }
+
+  if (surface == nullptr) {
+    FML_LOG(ERROR) << "SoftwareSurfaceProducer: Could not acquire surface";
```

Comment:
nit: Do we need the class name in these logs given that software_surface_producer.cc should already be tagged on the logs?

Reply:
Done